src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -83,9 +83,439 @@ module clsm_ensupd_read_obs
+ ! step 2: Clear Index (CI): CI > 20% is considered
+ !
+
+ !declariations of hdf functions
+ integer:: hopen, vfstart, vsfatch, vsqfnelt, vsfseek, vsfsfld, vsfread
+ integer:: vsfdtch, vfend, hclose
+
+ ! declarations of hdf-related parameters and variables
+ integer, dimension(N_files) :: file_id, vdata_id 
+ integer :: status, n_read, n_records, record_pos
+ integer, parameter :: num_dds_block = 0  ! only important for writing hdf
+ integer, parameter :: vdata_ref = 4
+ integer, parameter :: DFACC_READ     = 1 ! from hdf.inc
+ integer, parameter :: FULL_INTERLACE = 0 ! from hdf.inc
+ 
+ ! local variables
+ logical :: must_stop
+ integer, dimension(N_files) :: N_data_tmp
+
+ ! variables to define latitude and longitude    
+ integer :: i, j, k, k_off, ll, mm, kk
+ integer :: hh_num, time_index
+
+ integer, parameter :: bin_size = 0.05
+ integer, parameter :: XGRID = 3600
+ integer, parameter :: YGRID = 7200
+
+ real :: lat_ind(XGRID) = (/(ll, ll=0, 3600-1, 1)/)
+ real :: lon_ind(YGRID) = (/(mm, mm=0, 7200-1, 1)/)
+
+ !-------------------------------------------------------------
+ ! UTC | 0000    | 0300   | 0600  | 0900 | 1200 | 1500   | 1800    | 2100    
+ ! lon | 180 135 | 135 90 | 90 45 | 45 0 | 0 -45| -45 -90| -90 -135| -135 -180
+ !
+
+ real:: lon_subtime(9) = (/(kk, kk=180, -180, -45)/)
+     
+ integer, dimension(:), allocatable :: Snow_QA
+ integer, dimension(:), allocatable :: CI_Index, Cloud_index
+  
+ integer*2, dimension(:), allocatable :: tmpint2vec
+ real,      dimension(:), allocatable :: tmprealvec
+
+ character(len=*), parameter :: Iam = 'read_MOD10C1_hdf'
+ character(len=400) :: err_msg
+
+ lat_c = (90-bin_size/2)-bin_size*lat_ind
+ lon_c = (-180+bin_size/2)-bin_size*lon_ind
+
+ do i=1,7200
+ lat(3600*(i-1)+1:3600*i)= lat_c
+ lon(3600*(i-1)+1:3600*i)= lon_c(i)
+ end do
+
+  ! determine number of data to be read from each file
+    
+    do j=1,N_files
+       
+       ! open and "start" hdf file 
+       file_id(j) = hopen( fnames(j), DFACC_READ, num_dds_block )  
+       status = vfstart(file_id(j))
+         
+       ! select vdata block that contains fields of interest
+       vdata_id(j) = vsfatch(file_id(j), vdata_ref, 'r') 
+       
+       ! determine number of records in vdata
+       status = vsqfnelt(vdata_id(j), n_records)
+       N_data_tmp(j) = n_records
+       
+    end do
+
+    ! allocate pointers (must be deallocated outside this subroutine)
+    
+    must_stop = .false.
+    
+    if ( associated(lon) .or. associated(lat) .or. associated(MODIS_SCA) ) then
+       must_stop = .true.
+    end if
   
+    if (must_stop) then       
+    err_msg = 'output pointers must not be associated/allocated on input.'
+    call ldas_abort(LDAS_GENERIC_ERROR, Iam, err_msg)
+ 
+    end if
+
+    allocate(lon(N_data))
+    allocate(lat(N_data))
+    allocate(MODIS_SCA(N_data))
+    
+    allocate(CI_Index(N_data))
+    allocate(Snow_QA(N_data))
+    allocate(Cloud_Index(N_data))
```

Comment:
Has N_data been initialized?  It's "intent(out)" and looks like it's uninitialized at this point.